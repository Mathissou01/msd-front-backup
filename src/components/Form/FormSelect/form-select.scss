@use "src/styles/01-settings/settings.colors" as colors;

.c-FormSelect {
  display: flex;
  flex-direction: column;
  width: 100%;
}
