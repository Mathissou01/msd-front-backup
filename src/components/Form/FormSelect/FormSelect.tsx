import _ from "lodash";
import classNames from "classnames";
import { useFormContext } from "react-hook-form";
import { ErrorMessage } from "@hookform/error-message";
import React, { useEffect, useState } from "react";
import { IOptionWrapper } from "../FormMultiselect/FormMultiselect";
import FormLabel from "../FormLabel/FormLabel";
import CommonFormErrorText from "../../Common/CommonFormErrorText/CommonFormErrorText";
import "./form-select.scss";

interface IFormSelectProps<T> {
  name: string;
  label: string;
  secondaryLabel?: string;
  displayTransform?: (...args: Array<T>) => string;
  isRequired?: boolean;
  isDisabled?: boolean;
  options: Array<IOptionWrapper<T>>;
  optionKey?: keyof T & string;
  defaultValue?: T;
  noneSelectedLabel?: string;
  informationLabel?: string;
}

export default function FormSelect<T>({
  name,
  label,
  secondaryLabel,
  displayTransform,
  isRequired = false,
  isDisabled = false,
  options,
  optionKey,
  defaultValue,
  noneSelectedLabel,
  informationLabel,
}: IFormSelectProps<T>) {
  /* Static Data */
  noneSelectedLabel = noneSelectedLabel ?? "- Sélectionnez un élément -";
  const errorMessages = {
    required: "Ce champ est obligatoire",
  };

  /* Local Data */
  const [selectedIndex, setSelectedIndex] = useState<number>(-1);
  const {
    register,
    watch,
    setValue,
    formState: { isSubmitting, errors },
  } = useFormContext();
  const watchValue = watch(name);
  register(name, {
    required: { value: isRequired, message: errorMessages.required },
    value: defaultValue ?? null,
  });

  useEffect(() => {
    const matchIndex = options.findIndex(
      (wrapper) =>
        wrapper &&
        (optionKey
          ? wrapper.option?.[optionKey] === watchValue?.[optionKey]
          : wrapper.option === watchValue),
    );
    if (matchIndex !== selectedIndex) {
      setSelectedIndex(matchIndex);
    }
  }, [watchValue, options, optionKey, selectedIndex, setSelectedIndex]);

  function renderGroups() {
    return [...new Map(options.map((item) => [item["group"], item]))].map(
      (group, index) => {
        const groupOptions = options.filter(
          (option) => option.group === group[0],
        );
        return (
          groupOptions.length > 0 && (
            <optgroup
              className="o-SelectWrapper__OptGroup"
              key={`${name}_${index}_${group[0]}`}
              label={group[0]}
            >
              {renderOptions(groupOptions)}
            </optgroup>
          )
        );
      },
    );
  }

  function renderOptions(options: Array<IOptionWrapper<T>>) {
    return options.map((wrapper, index) => {
      return (
        wrapper.option && (
          <option
            className="o-SelectWrapper__Option"
            key={`${name}_${index}`}
            value={index}
          >
            {wrapper.option && !wrapper.label
              ? displayTransform
                ? displayTransform(wrapper.option)
                : optionKey
                ? wrapper.option[optionKey]?.toString()
                : wrapper.option.toString()
              : wrapper.label}
          </option>
        )
      );
    });
  }

  return (
    <div className="c-FormSelect">
      <FormLabel
        forId={name}
        label={label}
        isRequired={isRequired}
        secondaryLabel={secondaryLabel}
        informationLabel={informationLabel}
      />
      <div className="o-SelectWrapper">
        <select
          className={classNames("o-SelectWrapper__Select", {
            "o-SelectWrapper__Select_placeholder": selectedIndex < 0,
            "o-SelectWrapper__Select_invalid": _.get(errors, name),
          })}
          id={name}
          value={selectedIndex}
          onChange={(event) => {
            const index = Number.parseInt(event.target.value);
            setValue(name, options[index]?.option ?? null, {
              shouldValidate: true,
              shouldDirty: true,
            });
          }}
          disabled={isSubmitting || isDisabled}
          aria-invalid={!!_.get(errors, name)}
          aria-errormessage={`${name}_error`}
          data-testid="form-select"
        >
          <option disabled={isRequired} hidden={isRequired} value={-1}>
            {noneSelectedLabel}
          </option>
          {options.every((wrapper) => wrapper.group)
            ? renderGroups()
            : renderOptions(options)}
        </select>
      </div>
      <ErrorMessage
        errors={errors}
        name={name}
        render={({ message }: { message: string }) => (
          <CommonFormErrorText message={message} errorId={`${name}_error`} />
        )}
      />
    </div>
  );
}
