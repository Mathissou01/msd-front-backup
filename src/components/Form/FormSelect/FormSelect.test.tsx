import { render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import React, { ReactNode } from "react";
import FormSelect from "./FormSelect";

const mock = {
  name: "form-select",
  label: "form select",
  secondaryLabel: "secondary label",
  noneLabel: "none",
};

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return (
    <>
      <FormProvider {...formMethods}>{props.children}</FormProvider>
    </>
  );
};

describe("FormSelect", () => {
  it("renders", () => {
    const handleTransform = (data: { id: number; content: string }) => {
      return "-" + data.content;
    };
    const { container } = render(
      <Wrapper>
        <FormSelect<{ id: number; content: string }>
          name={mock.name}
          label={mock.label}
          secondaryLabel={mock.secondaryLabel}
          displayTransform={handleTransform}
          isRequired={false}
          isDisabled={false}
          options={[
            { option: { id: 1, content: "content 1" } },
            { option: { id: 2, content: "content 2" } },
          ]}
          optionKey={"id"}
          defaultValue={{ id: 1, content: "content 2" }}
          noneSelectedLabel={mock.noneLabel}
        />
      </Wrapper>,
    );
    expect(screen.getByTestId("form-select")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
