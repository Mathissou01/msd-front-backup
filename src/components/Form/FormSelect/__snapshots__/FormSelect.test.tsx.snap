// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect renders 1`] = `
<div>
  <div
    class="c-FormSelect"
  >
    <div
      class="c-FormLabel"
    >
      <label
        class="c-FormLabel__LabelWrapper"
        for="form-select"
      >
        <span
          class="c-FormLabel__Label"
        >
          form select
        </span>
      </label>
      <label
        class="c-FormLabel__LabelWrapper"
        for="form-select"
      >
        <span
          class="FormLabel__Secondary"
        >
          secondary label
        </span>
      </label>
    </div>
    <div
      class="o-SelectWrapper"
    >
      <select
        aria-errormessage="form-select_error"
        aria-invalid="false"
        class="o-SelectWrapper__Select o-SelectWrapper__Select_placeholder"
        data-testid="form-select"
        id="form-select"
      >
        <option
          value="-1"
        >
          none
        </option>
        <option
          class="o-SelectWrapper__Option"
          value="0"
        >
          -content 1
        </option>
        <option
          class="o-SelectWrapper__Option"
          value="1"
        >
          -content 2
        </option>
      </select>
    </div>
  </div>
</div>
`;
